test/files/general/shcounter.tla.gotests/shcounter_test.go

```diff
@@ -45,54 +48,112 @@ func getCounterValue(ctx *distsys.MPCalContext) (tla.TLAValue, error) {
 	return ctx.IFace().Read(arch, []tla.TLAValue{})
 }
 
-func TestShCounter(t *testing.T) {
-	numNodes := 5
-
+func makeContext(i int, receivers []*resources.TwoPCReceiver) *distsys.MPCalContext {
+	numNodes := len(receivers)
 	constants := []distsys.MPCalContextConfigFn{
 		distsys.DefineConstantValue("NUM_NODES", tla.MakeTLANumber(int32(numNodes))),
 	}
+	replicas := getReplicas(i, numNodes)
+	maker := resources.TwoPCArchetypeResourceMaker(
+		tla.MakeTLANumber(0),
+		getListenAddress(i),
+		replicas,
+		getArchetypeID(i),
+		func(receiver *resources.TwoPCReceiver) {
+			receivers[i] = receiver
+		},
+	)
+	return distsys.NewMPCalContext(tla.MakeTLANumber(int32(i)), ANode,
+		append(
+			constants,
+			distsys.EnsureArchetypeRefParam("cntr", maker),
+		)...,
+	)
+}
+
+func runTest(t *testing.T, numNodes int, injectFailures bool) {
+
+	var testMutex sync.Mutex
 
 	replicaCtxs := make([]*distsys.MPCalContext, numNodes)
+	receivers := make([]*resources.TwoPCReceiver, numNodes)
+	completed := make([]bool, numNodes)
 	errs := make(chan error, numNodes)
 
 	for i := 0; i < numNodes; i++ {
-		replicas := getReplicas(i, numNodes)
-		maker := resources.TwoPCArchetypeResourceMaker(
-			tla.MakeTLANumber(0),
-			getListenAddress(i),
-			replicas,
-			getArchetypeID(i),
-		)
-		ctx := distsys.NewMPCalContext(tla.MakeTLANumber(int32(i)), ANode,
-			append(
-				constants,
-				distsys.EnsureArchetypeRefParam("cntr", maker),
-			)...,
-		)
+		ctx := makeContext(i, receivers)
 		replicaCtxs[i] = ctx
+
+		ii := i
 		go func() {
-			errs <- runArchetype(ctx.Run)
+			runErr := runArchetype(ctx.Run)
+			testMutex.Lock()
+			completed[ii] = true
+			testMutex.Unlock()
+			if err := ctx.Close(); err != nil {
+				log.Println(err)
+			}
+			testMutex.Lock()
+			resources.CloseTwoPCReceiver(receivers[ii])
+			testMutex.Unlock()
+			log.Printf("Archetype %d has completed\n", ii)
+			errs <- runErr
 		}()
 	}
 
-	defer func() {
-		for _, ctx := range replicaCtxs {
-			if err := ctx.Close(); err != nil {
-				log.Println(err)
+	done := false
+	if injectFailures {
+		go func() {
+			for {
+				time.Sleep(1500 * time.Millisecond)
+				testMutex.Lock()
+				if done {
+					testMutex.Unlock()
+					break
+				}
+				indexToFail := rand.Intn(numNodes)
+				if completed[indexToFail] {
+					testMutex.Unlock()
+					log.Printf("Already done: %d\n", indexToFail)
+					continue
+				}
+				toFail := receivers[indexToFail]
+				log.Printf("Simulate failure on node %d\n", indexToFail)
+				testMutex.Unlock()
+				resources.SimulateTwoPCFailure(toFail)
```

Comment:
I have two questions about this test with failure injection. 

First, in my mind `shcounter` system is not fault-tolerant. Since processes wait until the value of counter becomes equal to the number of nodes. So if a node fails, there would be a deadlock.

Second, it's possible to terminate a context by calling `ctx.Close()`. I didn't understand why we're not using that here.

Reply:
TODO: remove failure injection